# Data Quality Assessment

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Analysing first sheet

In [2]:
# this sheet contains multiple sheets so pass the sheet name argument to read excel 
df = pd.read_excel(r"../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='Transactions')
df.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                     transaction_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1   Unnamed: 2           Unnamed: 3    Unnamed: 4    Unnamed: 5  \
0  product_id  customer_id     transaction_date  online_order  order_status   
1           2         2950  2017-02-25 00:00:00         False      Approved   
2           3         3120  2017-05-21 00:00:00          True      Approved   
3          37          402  2017-10-16 00:00:00         False      Approved   
4          88         3135  2017-08-31 00:00:00         False      Approved   

       Unnamed: 6    Unnamed: 7     Unnamed: 8    Unnamed: 9  ... Unnamed: 16  \
0           brand  product_line  product_class  product_size  ...         NaN   
1           Solex      Standard         medium        medium  ...         NaN   
2   Trek Bicycles      Standard         medium         large  ...         NaN   
3      OHM Cycles      Standard            low        medium  ...         NaN   
4  Norco Bicycles      Standard         medium        medium  ...         NaN   

  Unnamed: 17 Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  \
0         NaN         NaN          NaN          NaN          NaN          NaN   
1         NaN         NaN          NaN          NaN          NaN          NaN   
2         NaN         NaN          NaN          NaN          NaN          NaN   
3         NaN         NaN          NaN          NaN          NaN          NaN   
4         NaN         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  

[5 rows x 26 columns]

###  Cleaning the header row ,index column etc

In [3]:
# first should be column names for the dataset
col = df.loc[0]
df = df[1:]
df.rename(columns=col,inplace=True)
df.head()

transaction_id product_id customer_id     transaction_date online_order  \
1              1          2        2950  2017-02-25 00:00:00        False   
2              2          3        3120  2017-05-21 00:00:00         True   
3              3         37         402  2017-10-16 00:00:00        False   
4              4         88        3135  2017-08-31 00:00:00        False   
5              5         78         787  2017-10-01 00:00:00         True   

  order_status           brand product_line product_class product_size  ...  \
1     Approved           Solex     Standard        medium       medium  ...   
2     Approved   Trek Bicycles     Standard        medium        large  ...   
3     Approved      OHM Cycles     Standard           low       medium  ...   
4     Approved  Norco Bicycles     Standard        medium       medium  ...   
5     Approved  Giant Bicycles     Standard        medium        large  ...   

   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 26 columns]

In [4]:
# out of 26 columns only 13 columns were useful so remove 
df = df.iloc[:,:13]
df.head()

transaction_id product_id customer_id     transaction_date online_order  \
1              1          2        2950  2017-02-25 00:00:00        False   
2              2          3        3120  2017-05-21 00:00:00         True   
3              3         37         402  2017-10-16 00:00:00        False   
4              4         88        3135  2017-08-31 00:00:00        False   
5              5         78         787  2017-10-01 00:00:00         True   

  order_status           brand product_line product_class product_size  \
1     Approved           Solex     Standard        medium       medium   
2     Approved   Trek Bicycles     Standard        medium        large   
3     Approved      OHM Cycles     Standard           low       medium   
4     Approved  Norco Bicycles     Standard        medium       medium   
5     Approved  Giant Bicycles     Standard        medium        large   

  list_price standard_cost product_first_sold_date  
1      71.49         53.62                   41245  
2    2091.47        388.92                   41701  
3    1793.43        248.82                   36361  
4    1198.46         381.1                   36145  
5     1765.3        709.48                   42226

In [5]:
# setting transaction id as index
df.set_index('transaction_id',inplace = True)
df.head()

product_id customer_id     transaction_date online_order  \
transaction_id                                                            
1                       2        2950  2017-02-25 00:00:00        False   
2                       3        3120  2017-05-21 00:00:00         True   
3                      37         402  2017-10-16 00:00:00        False   
4                      88        3135  2017-08-31 00:00:00        False   
5                      78         787  2017-10-01 00:00:00         True   

               order_status           brand product_line product_class  \
transaction_id                                                           
1                  Approved           Solex     Standard        medium   
2                  Approved   Trek Bicycles     Standard        medium   
3                  Approved      OHM Cycles     Standard           low   
4                  Approved  Norco Bicycles     Standard        medium   
5                  Approved  Giant Bicycles     Standard        medium   

               product_size list_price standard_cost product_first_sold_date  
transaction_id                                                                
1                    medium      71.49         53.62                   41245  
2                     large    2091.47        388.92                   41701  
3                    medium    1793.43        248.82                   36361  
4                    medium    1198.46         381.1                   36145  
5                     large     1765.3        709.48                   42226

### Applying transformation of columns

In [6]:
df.dtypes

product_id                 object
customer_id                object
transaction_date           object
online_order               object
order_status               object
brand                      object
product_line               object
product_class              object
product_size               object
list_price                 object
standard_cost              object
product_first_sold_date    object
dtype: object

#### converting date time columns 
making changes in timestamps and other date columns

In [7]:
# we see transaction_date column as object dtype, we will change it to datetime dtype
df.transaction_date = pd.to_datetime(df.transaction_date)

#product first_sold_date is a timestamp in object dtype
df.product_first_sold_date = pd.to_datetime(df.product_first_sold_date).dt.date

In [8]:
df = df.astype({'product_id':int,"customer_id":int,'list_price':float,"standard_cost":float,'online_order':bool})

In [9]:
df.dtypes

product_id                          int32
customer_id                         int32
transaction_date           datetime64[ns]
online_order                         bool
order_status                       object
brand                              object
product_line                       object
product_class                      object
product_size                       object
list_price                        float64
standard_cost                     float64
product_first_sold_date            object
dtype: object

In [10]:
df.head()

product_id  customer_id transaction_date  online_order  \
transaction_id                                                           
1                        2         2950       2017-02-25         False   
2                        3         3120       2017-05-21          True   
3                       37          402       2017-10-16         False   
4                       88         3135       2017-08-31         False   
5                       78          787       2017-10-01          True   

               order_status           brand product_line product_class  \
transaction_id                                                           
1                  Approved           Solex     Standard        medium   
2                  Approved   Trek Bicycles     Standard        medium   
3                  Approved      OHM Cycles     Standard           low   
4                  Approved  Norco Bicycles     Standard        medium   
5                  Approved  Giant Bicycles     Standard        medium   

               product_size  list_price  standard_cost product_first_sold_date  
transaction_id                                                                  
1                    medium       71.49          53.62              1970-01-01  
2                     large     2091.47         388.92              1970-01-01  
3                    medium     1793.43         248.82              1970-01-01  
4                    medium     1198.46         381.10              1970-01-01  
5                     large     1765.30         709.48              1970-01-01

In [11]:
df.describe()

product_id   customer_id    list_price  standard_cost
count  20000.00000  20000.000000  20000.000000   19803.000000
mean      45.36465   1738.246050   1107.829449     556.046951
std       30.75359   1011.951046    582.825242     405.955660
min        0.00000      1.000000     12.010000       7.210000
25%       18.00000    857.750000    575.270000     215.140000
50%       44.00000   1736.000000   1163.890000     507.580000
75%       72.00000   2613.000000   1635.300000     795.100000
max      100.00000   5034.000000   2091.470000    1759.850000

#### Dealing with null values

In [12]:
df.isnull().sum()

product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [13]:
df.online_order = df.online_order.map({True:1,False:0})
df.online_order = df.online_order.astype(int)

In [14]:
df.dtypes

product_id                          int32
customer_id                         int32
transaction_date           datetime64[ns]
online_order                        int32
order_status                       object
brand                              object
product_line                       object
product_class                      object
product_size                       object
list_price                        float64
standard_cost                     float64
product_first_sold_date            object
dtype: object

In [15]:
df.order_status.value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [16]:
df[['brand','order_status']].value_counts(normalize=True)

brand           order_status
Solex           Approved        0.212645
Giant Bicycles  Approved        0.165783
WeareA2B        Approved        0.164874
OHM Cycles      Approved        0.152300
Trek Bicycles   Approved        0.149725
Norco Bicycles  Approved        0.145685
Solex           Cancelled       0.002121
WeareA2B        Cancelled       0.001515
Giant Bicycles  Cancelled       0.001464
OHM Cycles      Cancelled       0.001363
Norco Bicycles  Cancelled       0.001262
Trek Bicycles   Cancelled       0.001262
dtype: float64

In [17]:
df[df.index.duplicated()]

Empty DataFrame
Columns: [product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [18]:
df.isnull().sum()

product_id                   0
customer_id                  0
transaction_date             0
online_order                 0
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

We can either remove all the 197 rows or fill them most repeated value of that column

In [19]:
df.product_line.value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

In [20]:
df[df.index.duplicated()]

Empty DataFrame
Columns: [product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [21]:
df.product_class.value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [22]:
df[df['product_class'].isna()]

product_id  customer_id transaction_date  online_order  \
transaction_id                                                           
137                      0          431       2017-09-23             0   
160                      0         3300       2017-08-27             0   
367                      0         1614       2017-03-10             0   
407                      0         2559       2017-06-14             1   
677                      0         2609       2017-07-02             0   
...                    ...          ...              ...           ...   
19341                    0          443       2017-12-26             1   
19384                    0         2407       2017-06-11             0   
19794                    0         2860       2017-01-13             0   
19860                    0         2468       2017-06-24             1   
19872                    0           61       2017-03-17             1   

               order_status brand product_line product_class product_size  \
transaction_id                                                              
137                Approved   NaN          NaN           NaN          NaN   
160                Approved   NaN          NaN           NaN          NaN   
367                Approved   NaN          NaN           NaN          NaN   
407                Approved   NaN          NaN           NaN          NaN   
677                Approved   NaN          NaN           NaN          NaN   
...                     ...   ...          ...           ...          ...   
19341              Approved   NaN          NaN           NaN          NaN   
19384              Approved   NaN          NaN           NaN          NaN   
19794              Approved   NaN          NaN           NaN          NaN   
19860              Approved   NaN          NaN           NaN          NaN   
19872              Approved   NaN          NaN           NaN          NaN   

                list_price  standard_cost product_first_sold_date  
transaction_id                                                     
137                1942.61            NaN                     NaT  
160                1656.86            NaN                     NaT  
367                 850.89            NaN                     NaT  
407                 710.59            NaN                     NaT  
677                1972.01            NaN                     NaT  
...                    ...            ...                     ...  
19341               744.54            NaN                     NaT  
19384              1098.18            NaN                     NaT  
19794               868.56            NaN                     NaT  
19860              1497.43            NaN                     NaT  
19872               867.92            NaN                     NaT  

[197 rows x 12 columns]

No duplicated columns are present in the the dataset so no need to delete any rows

## Analyzing the second sheet 

In [23]:
# this sheet contains multiple sheets so pass the sheet name argument to read excel 
df1 = pd.read_excel(r"../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='NewCustomerList')
df1.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                         first_name                                                                                                                              
1                                            Chickie                                                                                                                              
2                                              Morly                                                                                                                              
3                                            Ardelis                                                                                                                              
4                                             Lucine                                                                                                                              

  Unnamed: 1 Unnamed: 2                           Unnamed: 3  \
0  last_name     gender  past_3_years_bike_related_purchases   
1    Brister       Male                                   86   
2     Genery       Male                                   69   
3  Forrester     Female                                   10   
4      Stutt     Female                                   64   

            Unnamed: 4                  Unnamed: 5             Unnamed: 6  \
0                  DOB                   job_title  job_industry_category   
1           1957-07-12             General Manager          Manufacturing   
2           1970-03-22         Structural Engineer               Property   
3  1974-08-28 00:00:00      Senior Cost Accountant     Financial Services   
4           1979-01-28  Account Representative III          Manufacturing   

          Unnamed: 7          Unnamed: 8 Unnamed: 9  ... Unnamed: 13  \
0     wealth_segment  deceased_indicator   owns_car  ...       state   
1      Mass Customer                   N        Yes  ...         QLD   
2      Mass Customer                   N         No  ...         NSW   
3  Affluent Customer                   N         No  ...         VIC   
4  Affluent Customer                   N        Yes  ...         QLD   

  Unnamed: 14         Unnamed: 15 Unnamed: 16 Unnamed: 17 Unnamed: 18  \
0     country  property_valuation         NaN         NaN         NaN   
1   Australia                   6        0.56      0.7000      0.8750   
2   Australia                  11        0.89      0.8900      1.1125   
3   Australia                   5        1.01      1.0100      1.0100   
4   Australia                   1        0.87      1.0875      1.0875   

   Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  
0          NaN          NaN         Rank        Value  
1     0.743750          1.0            1      1.71875  
2     0.945625          1.0            1      1.71875  
3     1.010000          1.0            1      1.71875  
4     1.087500          4.0            4     1.703125  

[5 rows x 23 columns]

In [24]:
col_name = df1.iloc[0]
df1=df1.loc[1:]
df1.rename(columns = col_name,inplace = True)

In [25]:
df1.shape

(1000, 23)

In [26]:
df1.columns

Index([                         'first_name',
                                 'last_name',
                                    'gender',
       'past_3_years_bike_related_purchases',
                                       'DOB',
                                 'job_title',
                     'job_industry_category',
                            'wealth_segment',
                        'deceased_indicator',
                                  'owns_car',
                                    'tenure',
                                   'address',
                                  'postcode',
                                     'state',
                                   'country',
                        'property_valuation',
                                         nan,
                                         nan,
                                         nan,
                                         nan,
                                         nan,
                                  

In [27]:
pd.set_option("display.max_columns",23)

In [28]:
df1.head()

first_name  last_name  gender past_3_years_bike_related_purchases  \
1    Chickie    Brister    Male                                  86   
2      Morly     Genery    Male                                  69   
3    Ardelis  Forrester  Female                                  10   
4     Lucine      Stutt  Female                                  64   
5    Melinda     Hadlee  Female                                  34   

                   DOB                   job_title job_industry_category  \
1           1957-07-12             General Manager         Manufacturing   
2           1970-03-22         Structural Engineer              Property   
3  1974-08-28 00:00:00      Senior Cost Accountant    Financial Services   
4           1979-01-28  Account Representative III         Manufacturing   
5           1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car tenure              address  \
1      Mass Customer                  N      Yes     14     45 Shopko Center   
2      Mass Customer                  N       No     16    14 Mccormick Park   
3  Affluent Customer                  N       No     10  5 Colorado Crossing   
4  Affluent Customer                  N      Yes      5   207 Annamark Plaza   
5  Affluent Customer                  N       No     19    115 Montana Place   

  postcode state    country property_valuation   NaN     NaN     NaN  \
1     4500   QLD  Australia                  6  0.56  0.7000  0.8750   
2     2113   NSW  Australia                 11  0.89  0.8900  1.1125   
3     3505   VIC  Australia                  5  1.01  1.0100  1.0100   
4     4814   QLD  Australia                  1  0.87  1.0875  1.0875   
5     2093   NSW  Australia                  9  0.52  0.5200  0.6500   

        NaN  NaN Rank     Value  
1  0.743750  1.0    1   1.71875  
2  0.945625  1.0    1   1.71875  
3  1.010000  1.0    1   1.71875  
4  1.087500  4.0    4  1.703125  
5  0.650000  4.0    4  1.703125

In [29]:
#there are some nan columns in the dataset 
df1 = df1.loc[:,df1.columns.notnull()]

In [30]:
df1.shape

(1000, 18)

In [31]:
df1.columns[16:21]

Index(['Rank', 'Value'], dtype='object')

In [32]:
df1

first_name  last_name  gender past_3_years_bike_related_purchases  \
1       Chickie    Brister    Male                                  86   
2         Morly     Genery    Male                                  69   
3       Ardelis  Forrester  Female                                  10   
4        Lucine      Stutt  Female                                  64   
5       Melinda     Hadlee  Female                                  34   
...         ...        ...     ...                                 ...   
996   Ferdinand  Romanetti    Male                                  60   
997        Burk    Wortley    Male                                  22   
998    Melloney      Temby  Female                                  17   
999      Dickie    Cubbini    Male                                  30   
1000      Sylas    Duffill    Male                                  56   

                      DOB                     job_title job_industry_category  \
1              1957-07-12               General Manager         Manufacturing   
2              1970-03-22           Structural Engineer              Property   
3     1974-08-28 00:00:00        Senior Cost Accountant    Financial Services   
4              1979-01-28    Account Representative III         Manufacturing   
5              1965-09-21             Financial Analyst    Financial Services   
...                   ...                           ...                   ...   
996            1959-10-07                     Paralegal    Financial Services   
997            2001-10-17        Senior Sales Associate                Health   
998            1954-10-05  Budget/Accounting Analyst IV    Financial Services   
999            1952-12-17             Financial Advisor    Financial Services   
1000           1955-10-02           Staff Accountant IV              Property   

         wealth_segment deceased_indicator owns_car tenure  \
1         Mass Customer                  N      Yes     14   
2         Mass Customer                  N       No     16   
3     Affluent Customer                  N       No     10   
4     Affluent Customer                  N      Yes      5   
5     Affluent Customer                  N       No     19   
...                 ...                ...      ...    ...   
996   Affluent Customer                  N       No      9   
997       Mass Customer                  N       No      6   
998   Affluent Customer                  N      Yes     15   
999       Mass Customer                  N      Yes     19   
1000      Mass Customer                  N      Yes     14   

                       address postcode state    country property_valuation  \
1             45 Shopko Center     4500   QLD  Australia                  6   
2            14 Mccormick Park     2113   NSW  Australia                 11   
3          5 Colorado Crossing     3505   VIC  Australia                  5   
4           207 Annamark Plaza     4814   QLD  Australia                  1   
5            115 Montana Place     2093   NSW  Australia                  9   
...                        ...      ...   ...        ...                ...   
996                2 Sloan Way     2200   NSW  Australia                  7   
997          04 Union Crossing     2196   NSW  Australia                 10   
998   33475 Fair Oaks Junction     4702   QLD  Australia                  2   
999         57666 Victoria Way     4215   QLD  Australia                  2   
1000        21875 Grover Drive     2010   NSW  Australia                  9   

      Rank     Value  
1        1   1.71875  
2        1   1.71875  
3        1   1.71875  
4        4  1.703125  
5        4  1.703125  
...    ...       ...  
996    996     0.374  
997    997     0.357  
998    997     0.357  
999    997     0.357  
1000  1000      0.34  

[1000 rows x 18 columns]

In [33]:
df1.job_industry_category.unique()

array(['Manufacturing', 'Property', 'Financial Services', 'Entertainment',
       'Retail', 'IT', 'Telecommunications', 'Health', nan, 'Argiculture'],
      dtype=object)

In [34]:
df1.dtypes

first_name                             object
last_name                              object
gender                                 object
past_3_years_bike_related_purchases    object
DOB                                    object
job_title                              object
job_industry_category                  object
wealth_segment                         object
deceased_indicator                     object
owns_car                               object
tenure                                 object
address                                object
postcode                               object
state                                  object
country                                object
property_valuation                     object
Rank                                   object
Value                                  object
dtype: object

In [35]:
df1.rename({"past_3_years_bike_related_purchases":'other_purchases',"job_industry_category":"job_industry"},axis = 1,inplace=True)

In [36]:
df1 = df1.astype({"other_purchases":int,"tenure":int,"property_valuation" :int,"Rank":int,"Value":float})

In [37]:
df1.Value = df1.Value.round(3)

In [38]:
df1.head()

first_name  last_name  gender  other_purchases                  DOB  \
1    Chickie    Brister    Male               86           1957-07-12   
2      Morly     Genery    Male               69           1970-03-22   
3    Ardelis  Forrester  Female               10  1974-08-28 00:00:00   
4     Lucine      Stutt  Female               64           1979-01-28   
5    Melinda     Hadlee  Female               34           1965-09-21   

                    job_title        job_industry     wealth_segment  \
1             General Manager       Manufacturing      Mass Customer   
2         Structural Engineer            Property      Mass Customer   
3      Senior Cost Accountant  Financial Services  Affluent Customer   
4  Account Representative III       Manufacturing  Affluent Customer   
5           Financial Analyst  Financial Services  Affluent Customer   

  deceased_indicator owns_car  tenure              address postcode state  \
1                  N      Yes      14     45 Shopko Center     4500   QLD   
2                  N       No      16    14 Mccormick Park     2113   NSW   
3                  N       No      10  5 Colorado Crossing     3505   VIC   
4                  N      Yes       5   207 Annamark Plaza     4814   QLD   
5                  N       No      19    115 Montana Place     2093   NSW   

     country  property_valuation  Rank  Value  
1  Australia                   6     1  1.719  
2  Australia                  11     1  1.719  
3  Australia                   5     1  1.719  
4  Australia                   1     4  1.703  
5  Australia                   9     4  1.703

In [39]:
df1.isnull().sum()

first_name              0
last_name              29
gender                  0
other_purchases         0
DOB                    17
job_title             106
job_industry          165
wealth_segment          0
deceased_indicator      0
owns_car                0
tenure                  0
address                 0
postcode                0
state                   0
country                 0
property_valuation      0
Rank                    0
Value                   0
dtype: int64

In [40]:
#if person dont have last name fill it with "." or we can fill it with first name
df1.last_name = df1.last_name.fillna(".")

In [41]:
df1.job_title.value_counts()

Associate Professor             15
Environmental Tech              14
Software Consultant             14
Chief Design Engineer           13
Assistant Media Planner         12
                                ..
Safety Technician IV             1
Administrative Officer           1
Budget/Accounting Analyst II     1
Statistician IV                  1
Systems Administrator III        1
Name: job_title, Length: 184, dtype: int64

In [42]:
df1.DOB = pd.to_datetime(df1.DOB)

In [43]:
pd.to_datetime(df1.DOB).dt.year.mode()

0    1974.0
dtype: float64

In [44]:
df1.DOB.fillna('1974-01-01',inplace = True)

In [74]:
df1["year"] = df1.DOB.apply(lambda x : str(x).split("-")[0])

In [76]:
df1.year.sort_values()

192    1938
407    1938
886    1938
978    1938
586    1938
       ... 
142    2001
997    2001
570    2001
764    2002
397    2002
Name: year, Length: 751, dtype: object

In [45]:
df1.dropna(inplace=True)

In [46]:
df1.isna().sum()

first_name            0
last_name             0
gender                0
other_purchases       0
DOB                   0
job_title             0
job_industry          0
wealth_segment        0
deceased_indicator    0
owns_car              0
tenure                0
address               0
postcode              0
state                 0
country               0
property_valuation    0
Rank                  0
Value                 0
dtype: int64

## Analyzing 3rd sheet

In [47]:
df2 = pd.read_excel("../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='CustomerDemographic')

In [48]:
df2.head()

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                        customer_id                                                                                                                              
1                                                  1                                                                                                                              
2                                                  2                                                                                                                              
3                                                  3                                                                                                                              
4                                                  4                                                                                                                              

   Unnamed: 1 Unnamed: 2 Unnamed: 3                           Unnamed: 4  \
0  first_name  last_name     gender  past_3_years_bike_related_purchases   
1     Laraine  Medendorp          F                                   93   
2         Eli    Bockman       Male                                   81   
3       Arlin     Dearle       Male                                   61   
4      Talbot        NaN       Male                                   33   

            Unnamed: 5              Unnamed: 6             Unnamed: 7  \
0                  DOB               job_title  job_industry_category   
1  1953-10-12 00:00:00     Executive Secretary                 Health   
2  1980-12-16 00:00:00  Administrative Officer     Financial Services   
3  1954-01-20 00:00:00      Recruiting Manager               Property   
4  1961-10-03 00:00:00                     NaN                     IT   

       Unnamed: 8          Unnamed: 9  \
0  wealth_segment  deceased_indicator   
1   Mass Customer                   N   
2   Mass Customer                   N   
3   Mass Customer                   N   
4   Mass Customer                   N   

                                         Unnamed: 10  ... Unnamed: 15  \
0                                            default  ...         NaN   
1                                                 "'  ...         NaN   
2                       <script>alert('hi')</script>  ...         NaN   
3                                2018-02-01 00:00:00  ...         NaN   
4  () { _; } >_[$($())] { touch /tmp/blns.shellsh...  ...         NaN   

  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  \
0         NaN          NaN          NaN          NaN          NaN   
1         NaN          NaN          NaN          NaN          NaN   
2         NaN          NaN          NaN          NaN          NaN   
3         NaN          NaN          NaN          NaN          NaN   
4         NaN          NaN          NaN          NaN          NaN   

   Unnamed: 21  Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN          NaN  

[5 rows x 26 columns]

In [49]:
col_name = df2.loc[0]
df2 = df2[1:]
df2.rename(columns=col_name,inplace = True)

In [50]:
df2.set_index('customer_id',inplace = True)

In [51]:
df2.head()

first_name  last_name  gender  \
customer_id                                      
1                   Laraine  Medendorp       F   
2                       Eli    Bockman    Male   
3                     Arlin     Dearle    Male   
4                    Talbot        NaN    Male   
5            Sheila-kathryn     Calton  Female   

            past_3_years_bike_related_purchases                  DOB  \
customer_id                                                            
1                                            93  1953-10-12 00:00:00   
2                                            81  1980-12-16 00:00:00   
3                                            61  1954-01-20 00:00:00   
4                                            33  1961-10-03 00:00:00   
5                                            56  1977-05-13 00:00:00   

                          job_title job_industry_category     wealth_segment  \
customer_id                                                                    
1               Executive Secretary                Health      Mass Customer   
2            Administrative Officer    Financial Services      Mass Customer   
3                Recruiting Manager              Property      Mass Customer   
4                               NaN                    IT      Mass Customer   
5                     Senior Editor                   NaN  Affluent Customer   

            deceased_indicator  \
customer_id                      
1                            N   
2                            N   
3                            N   
4                            N   
5                            N   

                                                       default owns_car  ...  \
customer_id                                                              ...   
1                                                           "'      Yes  ...   
2                                 <script>alert('hi')</script>      Yes  ...   
3                                          2018-02-01 00:00:00      Yes  ...   
4            () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No  ...   
5                                                          NIL      Yes  ...   

             NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
customer_id                                                         
1            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
2            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
3            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5            NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 25 columns]

In [52]:
df2 = df2.loc[:,df2.columns.notnull()]
df2.head()

first_name  last_name  gender  \
customer_id                                      
1                   Laraine  Medendorp       F   
2                       Eli    Bockman    Male   
3                     Arlin     Dearle    Male   
4                    Talbot        NaN    Male   
5            Sheila-kathryn     Calton  Female   

            past_3_years_bike_related_purchases                  DOB  \
customer_id                                                            
1                                            93  1953-10-12 00:00:00   
2                                            81  1980-12-16 00:00:00   
3                                            61  1954-01-20 00:00:00   
4                                            33  1961-10-03 00:00:00   
5                                            56  1977-05-13 00:00:00   

                          job_title job_industry_category     wealth_segment  \
customer_id                                                                    
1               Executive Secretary                Health      Mass Customer   
2            Administrative Officer    Financial Services      Mass Customer   
3                Recruiting Manager              Property      Mass Customer   
4                               NaN                    IT      Mass Customer   
5                     Senior Editor                   NaN  Affluent Customer   

            deceased_indicator  \
customer_id                      
1                            N   
2                            N   
3                            N   
4                            N   
5                            N   

                                                       default owns_car tenure  
customer_id                                                                     
1                                                           "'      Yes     11  
2                                 <script>alert('hi')</script>      Yes     16  
3                                          2018-02-01 00:00:00      Yes     15  
4            () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No      7  
5                                                          NIL      Yes      8

In [53]:
df2.drop("default",axis = 1,inplace = True)

In [54]:
df2.isna().sum()

first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [55]:
df2.rename({"past_3_years_bike_related_purchases":"other_purchases","job_industry_category":"job_industry"},axis = 1,inplace = True)

In [56]:
df2.head()

first_name  last_name  gender other_purchases  \
customer_id                                                      
1                   Laraine  Medendorp       F              93   
2                       Eli    Bockman    Male              81   
3                     Arlin     Dearle    Male              61   
4                    Talbot        NaN    Male              33   
5            Sheila-kathryn     Calton  Female              56   

                             DOB               job_title        job_industry  \
customer_id                                                                    
1            1953-10-12 00:00:00     Executive Secretary              Health   
2            1980-12-16 00:00:00  Administrative Officer  Financial Services   
3            1954-01-20 00:00:00      Recruiting Manager            Property   
4            1961-10-03 00:00:00                     NaN                  IT   
5            1977-05-13 00:00:00           Senior Editor                 NaN   

                wealth_segment deceased_indicator owns_car tenure  
customer_id                                                        
1                Mass Customer                  N      Yes     11  
2                Mass Customer                  N      Yes     16  
3                Mass Customer                  N      Yes     15  
4                Mass Customer                  N       No      7  
5            Affluent Customer                  N      Yes      8

In [57]:
df2.gender.value_counts()

Female    2037
Male      1872
U           88
Femal        1
M            1
F            1
Name: gender, dtype: int64

In [58]:
df2.gender.replace({"Female":"F","Male":"M","Femal":"F","U":"Not Specified"},inplace=True)

In [59]:
df2.gender.value_counts()

F                2039
M                1873
Not Specified      88
Name: gender, dtype: int64

In [60]:
df2.DOB = pd.to_datetime(df2.DOB).dt.date

In [61]:
pd.to_datetime(df2.DOB).dt.year.mode()

0    1978.0
dtype: float64

In [62]:
df2.DOB.fillna('1978-01-01',inplace = True)

In [63]:
df2.DOB = pd.to_datetime(df2.DOB).dt.date

In [69]:
df2["year"] = df2.DOB.apply(lambda x : str(x).split("-")[0])

In [70]:
df2.year.sort_values()

customer_id
34      1843
720     1931
1092    1935
3410    1940
2413    1943
        ... 
422     2002
2296    2002
2858    2002
3435    2002
1888    2002
Name: DOB, Length: 4000, dtype: object

In [71]:
#strange error in DOB column with 1843 
df2.drop(34,axis = 0,inplace=True)

KeyError: '[34] not found in axis'

In [72]:
df2.columns

Index(['first_name', 'last_name', 'gender', 'other_purchases', 'DOB',
       'job_title', 'job_industry', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'year'],
      dtype='object')

In [73]:
df2.drop(['year'],axis=1)

first_name  last_name         gender other_purchases  \
customer_id                                                             
1                   Laraine  Medendorp              F              93   
2                       Eli    Bockman              M              81   
3                     Arlin     Dearle              M              61   
4                    Talbot        NaN              M              33   
5            Sheila-kathryn     Calton              F              56   
...                     ...        ...            ...             ...   
3996                Rosalia   Halgarth              F               8   
3997                 Blanch     Nisuis              F              87   
3998                 Sarene    Woolley  Not Specified              60   
3999              Patrizius        NaN              M              11   
4000                  Kippy    Oldland              M              76   

                    DOB               job_title        job_industry  \
customer_id                                                           
1            1953-10-12     Executive Secretary              Health   
2            1980-12-16  Administrative Officer  Financial Services   
3            1954-01-20      Recruiting Manager            Property   
4            1961-10-03                     NaN                  IT   
5            1977-05-13           Senior Editor                 NaN   
...                 ...                     ...                 ...   
3996         1975-08-09   VP Product Management              Health   
3997         2001-07-13         Statistician II       Manufacturing   
3998         1978-01-01       Assistant Manager                  IT   
3999         1973-10-24                     NaN       Manufacturing   
4000         1991-11-05    Software Engineer IV                 NaN   

                wealth_segment deceased_indicator owns_car tenure  
customer_id                                                        
1                Mass Customer                  N      Yes     11  
2                Mass Customer                  N      Yes     16  
3                Mass Customer                  N      Yes     15  
4                Mass Customer                  N       No      7  
5            Affluent Customer                  N      Yes      8  
...                        ...                ...      ...    ...  
3996             Mass Customer                  N       No     19  
3997            High Net Worth                  N      Yes      1  
3998            High Net Worth                  N       No    NaN  
3999         Affluent Customer                  N      Yes     10  
4000         Affluent Customer                  N       No     11  

[3999 rows x 11 columns]

In [ ]:
df2.last_name.fillna('.',inplace=True)

In [ ]:
df2.isna().sum()

In [ ]:
df2.tenure.fillna(np.mean(df2.tenure),inplace = True)

In [ ]:
np.mean(df2.tenure)

In [ ]:
df2.drop_duplicates(inplace=True)

In [ ]:
df2.dtypes

In [ ]:
df2 = df2.astype({"other_purchases":int,"tenure":int})

In [ ]:
df2.head()

###  Analyzing the third sheet

In [ ]:
df3 = pd.read_excel("../KPMG_VI_New_raw_data_update_final.xlsx",sheet_name='CustomerAddress')
df3

In [ ]:
col_name = df3.iloc[0]
df3 = df3[1:]
df3.rename(columns=col_name,inplace = True)


In [ ]:
#removing the nan values
df3 = df3.loc[:,df3.columns.notnull()]

In [ ]:
df3.head()

In [ ]:
df3.drop_duplicates()

In [ ]:
df3.customer_id.unique() == np.arange(1,4000)

In [ ]:
#since all values in country column is Australia it doesnt provide any info so we can delete or rename dataframe with australia dataset

In [ ]:
df3.state.value_counts()

In [ ]:
df3.drop_duplicates(inplace=True)
#no duplicates to remove

In [ ]:
df3.property_valuation.value_counts()